### DISCUSSION TITLE
Why does V compile so fast?

### DISCUSSION THREAD

**Discussion by JasonGlazer** [vlang/v#13555] [General]
The compile speeds on V are really impressive. Can someone describe the approach that makes is so quick? Is it because it targets C?

**Comment by JalonSolov** [comment]
The C compiler is often the slowest part of the whole process.  V itself is _very_ fast up to creating the C code, then it calls C to deal with it.

One thing that makes it faster is that V creates a single `.c` file - no headers, no extra `.c` files to link with, etc.  This is the fastest thing a C compiler can handle.

Also, by default, V invokes the `tcc` compiler, which is itself very fast, but mainly because it doesn't do _any_ optimizations.  You can see the difference by just adding `-prod` to your V command line (as in, instead of `v file.v` use `v -prod file.v`).  This causes V to call msvc (on Windows, of course), gcc, or clang, whichever it finds, and also passes a flag to tell the C compiler to turn on full optimizations.

**Reply by JasonGlazer** [reply]
So why is V very fast at creating C code? What algorithms and tricks does it use to be so fast?

**Reply by medvednikov** [reply]
Just keeping it simple. No over-engineering, simple code, focus on efficiency.

**Reply by igotfr** [reply]
> So why is V very fast at creating C code? What algorithms and tricks does it use to be so fast?

The entire V compiler is here: https://github.com/vlang/v/tree/master/vlib/v

One thing I don't understand is: the V compiler is made in V, it's normal and can be made in any language, okay. But where is the initial V compiler code that was made in C or Golang?

**Reply by medvednikov** [reply]
It was written in Go. It's not needed for bootstrapping since V compiles to C, so there's no need in it.

It may be released in the future just for history.

**Comment by JalonSolov** [comment]
The original Golang version was never made public.  It only existed on Alex's machine until V could compile itself.  He may have deleted the Golang version after that, since it wasn't needed any longer.
